platforms/allwinner-d1/boards/lichee-rv/src/main.rs

```diff
@@ -0,0 +1,148 @@
+    // both purposes, as this will likely play better with sleep power usage.
+    let Timers { mut timer0, mut timer1 } = Timers::new(p.TIMER);
+
+    let k = initialize_kernel().unwrap();
+
+    k.initialize(async move {
+        let mut i = 0;
+        loop {
+            let ticks = WFI_CT.load(Ordering::Relaxed);
+            write!(&mut uart, "Hello, world! {i} - {ticks}\r\n").ok();
+            i += 1;
+            k.sleep(Duration::from_millis(250)).await;
+        }
+    }).unwrap();
+
+    timer0.set_prescaler(TimerPrescaler::P8); // 24M / 8:  3.00M ticks/s
+    timer1.set_prescaler(TimerPrescaler::P8);
+    timer0.set_mode(TimerMode::PERIODIC);
+    timer1.set_mode(TimerMode::SINGLE_COUNTING);
+    let _ = timer0.get_and_clear_interrupt();
+    let _ = timer1.get_and_clear_interrupt();
+
+    unsafe {
+        riscv::interrupt::enable();
+        riscv::register::mie::set_mext();
+    }
+
+
+    let plic = Plic::new(p.PLIC);
+
+    unsafe {
+        plic.register(Interrupt::TIMER1, timer1_int);
+    }
+
+    timer0.start_counter(0xFFFF_FFFF);
+
+    loop {
+        // Tick the scheduler
+        let start = timer0.current_value();
+        let tick = k.tick();
+
+        // Timer is downcounting
+        let elapsed = start.wrapping_sub(timer0.current_value());
+        let turn = k.timer().force_advance_ticks(elapsed.into());
+
+        // If there is nothing else scheduled, sleep for some amount of time
+        if !tick.has_remaining {
+            let wfi_start = timer0.current_value();
+
+            // TODO(AJM): Sometimes there is no "next" in the timer wheel, even though there should
+            // be. Don't take lack of timer wheel presence as the ONLY heuristic of whether we
+            // should just wait for SOME interrupt to occur. For now, force a max sleep of 100ms
+            // which is still probably wrong.
+            let amount = turn.ticks_to_next_deadline()
+                .unwrap_or(100 * 1000 * 3); // 3 ticks per us, 1000 us per ms, 100ms sleep
+
+            // Don't sleep for too long until james figures out wrapping timers
+            let amount = amount.min(0x4000_0000) as u32;
+            let _ = timer1.get_and_clear_interrupt();
+            unsafe {
+                plic.activate(Interrupt::TIMER1, Priority::P1).unwrap();
+            }
+            timer1.set_interrupt_en(true);
+            timer1.start_counter(amount);
+
+            unsafe {
+                WFI_CT.fetch_add(1, Ordering::Relaxed);
+                riscv::asm::wfi();
+            }
+            // Disable the timer interrupt in case that wasn't what woke us up
+            plic.deactivate(Interrupt::TIMER1).unwrap();
+            timer1.set_interrupt_en(false);
+            timer1.stop();
+
+            // Account for time slept
+            let elapsed = wfi_start.wrapping_sub(timer0.current_value());
+            let _turn = k.timer().force_advance_ticks(elapsed.into());
```

Comment:
we should advance the timer wheel on every iteration of the loop even when there *are* tasks remaining in the run queue, so i think this should be moved out of the if. otherwise, expiring timers will not be notified when the system is busy. 

we should probably move advancing the timer wheel to *before* the "wait for interrupt" phase, and only wait for an interrupt if the `Turn` struct indicates that no timers completed. in general i think the overall structure should probably look closer to the run loop in melpomene.

Reply:
UPDATE: oh, i had missed that there was already a call to advance the wheel before the if. never mind, this looks right!